Implement Physical Light Units in Vulkan Renderers

clayjohn opened the pull request:
This closes: https://github.com/godotengine/godot-proposals/issues/4257

This allows light sources to be specified in physical light units in addition to the regular energy multiplier. In order to avoid loss of precision at high values, brightness values are premultiplied by an exposure normalization value.

In support of Physical Light Units this PR also renames CameraEffects to CameraAttributes.

This PR fully implements Physical Light Units and I believe it is mostly working as intended. I just need to polish the code and search for untested edge cases. I will update this description as I polish the feature.

Still TODO:
~1. clean up, polish, and rebase~
~2. Physical Lens properties~
~3. Implement exposure settings in GLES3 and Vulkan Mobile~
~4. Documentation~

clayjohn committed:
Implement Physical Light Units as an optional setting.

This allows light sources to be specified in physical light units in addition to the regular energy multiplier. In order to avoid loss of precision at high values, brightness values are premultiplied by an exposure normalization value.

In support of Physical Light Units this PR also renames CameraEffects to CameraAttributes.

jcostello commented:
Any change in visual or is only a matter of unit definition?

clayjohn commented:
> Any change in visual or is only a matter of unit definition?

Default visuals are the same. But now lights can be much brighter without breaking rendering and lights can optionally specify color temperature to modify their color (in addition to the regular color property)

jcostello commented:
Im trying to test this PR. So far I see everything shining white. I will keep testing it later. 

BTW, Intensity is only in lux. Is other unit going to be implemented? Can the temperature be a slider with a temperature color hint?

clayjohn commented:
> Im trying to test this PR. So far I see everything shining white. I will keep testing it later.

You will need to add a CameraAttributes resource to your WorldEnvironment using it you can set your exposure settings. By default a brightness of 100000 lux will make your image over-exposed. The default settings for CameraAttributesPhysical should work fine.

> BTW, Intensity is only in lux. Is other unit going to be implemented? 

I have no plans to add another unit. For directional lights Lux is the only unit that makes sense. Omnilights and Spotlights are in Lumens, I have no plans to add another unit, but we could add candela's if people really need it.

Other light sources (Sky, emission etc.) are in nits already. For these light sources, only nits really makes sense

> Can the temperature be a slider with a temperature color hint?

Sure. I was going to make a proposal for that after this PR is merged. I am not good with editor/ui stuff so I will leave that for someone else. What would be nice is just having the color spectrum with a slider that moves across it

jcostello commented:
Testing it out. A few things to mention:

Lightmaps dont work, even with the camera attribute set on it.
The speed attribute in the auto exposure setting influences the exposure itself.

reduz commented:
@clayjohn I would recommend you to post some screenshots for people to appreciate your work on more detail.

clayjohn commented:
> Testing it out. A few things to mention:
> 
> Lightmaps dont work, even with the camera attribute set on it. The speed attribute in the auto exposure setting influences the exposure itself.

In my testing Lightmaps work fine, have you set CameraAttributes on both the Lightmap (before baking) and on the WorldEnvironment? Maybe there is some edge case I have not tested. If you have a simple MRP you could share that would be super helpful!

I haven't changed the auto exposure effect and I am unlikely to dig too deep into it in this PR as the size is already getting out of hand.

jcostello commented:
Im submitting MRP. I added a CameraAttributes to the environment and the lightmap before baking.

Lightmap disabled

![image](https://user-images.githubusercontent.com/1168582/182274107-80969ab1-67c4-484b-8a0f-7b5edc603135.png)

Lightmap enabled

![image](https://user-images.githubusercontent.com/1168582/182274153-25f9846c-0389-4ee7-85a2-b8ca5724b9a1.png)

jcostello commented:
MRP

[Lightmap.zip](https://github.com/godotengine/godot/files/9238414/Lightmap.zip)

clayjohn commented:
Some nice screenshots of SDFGI playing nicely with Physically Based Light units. 
![Screenshot from 2022-08-10 12-28-10](https://user-images.githubusercontent.com/16521339/184237137-1415eca6-1cd3-4746-9dec-25eb6d3f8193.png)
![Screenshot from 2022-08-10 12-27-38](https://user-images.githubusercontent.com/16521339/184237139-662dde6b-1e0d-4dec-ae3b-56d4368125d1.png)

clayjohn commented:
@jcostello I am running your MRP now. It looks like you used a CameraAttributesPractical with default exposure, but your DirectionalLight3D has an intensity of 100000 lux (also default). This causes the LightmapGI to be way over exposed. Swapping out the CameraAttributesPractical for a CameraAttributesPhysical (with default settings) makes your MRP look like this:

![Screenshot from 2022-08-12 11-58-13](https://user-images.githubusercontent.com/16521339/184416630-25c1f3a3-d5d4-4f1e-9bc7-b66ea2170ed3.png)

I will clarify this more in the documentation. But when using Physical Light Units you have to be very careful that you manage the dynamic range well with exposure settings. 

This is a big part of why we have avoided implementing them until now. When baking lighting, users are going to have to manually set CameraAttributes settings to avoid overexposing and overflowing colors, or underexposing and creating banding.

jcostello commented:
> @jcostello I am running your MRP now. It looks like you used a CameraAttributesPractical with default exposure, but your DirectionalLight3D has an intensity of 100000 lux (also default). This causes the LightmapGI to be way over exposed. Swapping out the CameraAttributesPractical for a CameraAttributesPhysical (with default settings) makes your MRP look like this:
> 
> ![Screenshot from 2022-08-12 11-58-13](https://user-images.githubusercontent.com/16521339/184416630-25c1f3a3-d5d4-4f1e-9bc7-b66ea2170ed3.png)
> 
> I will clarify this more in the documentation. But when using Physical Light Units you have to be very careful that you manage the dynamic range well with exposure settings.
> 
> This is a big part of why we have avoided implementing them until now. When baking lighting, users are going to have to manually set CameraAttributes settings to avoid overexposing and overflowing colors, or underexposing and creating banding.

Thanks. I tried this and worked :muscle:

clayjohn commented:
Now physically-based DoF is working. 

Here is an example with various settings and a telephoto lens:

_Aperture: 16, Focal length: 200, Focus distance: 1m_
![Screenshot from 2022-08-17 17-18-53](https://user-images.githubusercontent.com/16521339/185260200-08f73913-4760-4252-a23f-7a6bb4f668f9.png)

_Aperture: 16, Focal length: 200, Focus distance: 5m_
![Screenshot from 2022-08-17 17-19-04](https://user-images.githubusercontent.com/16521339/185260197-91ca0892-f508-430e-be6c-c65780f52399.png)

_Aperture: 16, Focal length: 200, Focus distance: 9m_
![Screenshot from 2022-08-17 17-19-11](https://user-images.githubusercontent.com/16521339/185260196-d4a89e9f-607b-4efe-b8fa-4960df78e22f.png)

_Aperture: 4, Focal length: 200, Focus distance: 5m_
![Screenshot from 2022-08-17 17-19-30](https://user-images.githubusercontent.com/16521339/185260188-5f2f5f9b-923f-4d18-a8e5-a228c91eb52e.png)
_Note: Shutter speed is also changed to compensate for the increase in aperture size_

Some of the settings are a little adhoc as I used the existing DoF shader instead of writing a stochastic one from scratch. I think the results are pretty convincing (we may just need to tweak the total blur amount if this isn't realistic).

tinmanjuggernaut commented:
I tested this PR. There's some good stuff, and some very broken stuff. It was getting too much to describe in text, so I made a video review:
https://youtu.be/CZm3WdFNUps (31min)

Summary points:
* Default environment sun extremely large
* Glow is strange
* DirectionalLight is effectively 100,000x too strong, which blows out all materials
* Blown out materials turn black instead of staying white
* Blown out materials hide gizmo controls
* Material previews are black (for basic white material)
* Default camera settings create a black environment. The numbers are fine defaults, the effects are inaccurate.
* Blown out objects create a shape larger than the actual object
* Focal length doesn't change field of view, and vica versa
* FOV is only partially locked out under Phys Camera. Either unlock it or completely lock it.
* Exposure settings should all be in the exposure group
* Recommend reordering settings: Focus Distance, Focal Length, Aperture, Shutter speed, Sensitivity, Multiplier.
* Auto exposure only sometimes works, doesn't override exposure settings
* Physical Lenses have different blur amounts for back focal plane and front focal plane depending on circumstances. Both the practical and physical cameras should have the option. The latter seems to have it already.
* The Physical Camera does seem to be fairly accurate, supporting separate front/back blur amounts. I need to do more testing once I can get set up in my game with more complex objects. However I have only one minor issue with it below. I think it's good enough for this PR. I'll do more testing once I can get my game working with it in GD4.
* The Physical Camera amount of blur is a bit weak. I was estimating the amount of blur on an object at 2m away is what 3m away should look like. So perhaps a 66.7% increase will achieve it. Your photos above are not representative of real photography. 200mm lens is a huge, telephoto lens. It's an odd choice to test with. An aperture of f/16 generally means everything is in focus. So I understand why you had to push the focal length all the way to 200 to get blur on f/16. 50mm lenses are considered normal, or the equivalent of how human vision works. Better testing numbers would be in the ranges of a normal 50mm f/1 to f/4, or a portrait 90mm f/1 to f/8. Using the DOF simulator to compare blur amounts as I did in the video is a reasonable method.

* Here's how I change fov and focal length in real time:
```
func set_focal_length(value: float) -> void:
		focal_length = value
		fov = focal2fov(focal_length)
		focus_at(focus_point)


var _set_wait: bool = false
func set_fov(value: float) -> void:
		fov = value
		focal_length = fov2focal(fov)
		print("set_fov %f %f" % [ value, focal_length ])
		if is_inside_tree() and not _set_wait:
			_set_wait = true
			await get_tree().create_timer(1.0).timeout	# Allow user to update for 1 sec before processing sub tasks
			_set_wait = false
			focus_at(focus_point)

static func focal2fov(focal: float) -> float:
	return atan(12.0/focal) * 114.5916	# Simplified atan(24/2/focal_length) * 180/PI * 2
	
static func fov2focal(p_fov: float) -> float:
	return 12.0/tan(p_fov/114.5916)
	
```

clayjohn commented:
@tinmanjuggernaut Thank you for your comments, most of them I can address as I finish the PR!

clayjohn commented:
@tinmanjuggernaut Thank you for your comments. I have addressed most of them in the most recent push and I will now watch the video before seeing if I can address the rest.

> * Default environment sun extremely large

Fixed

> * DirectionalLight is effectively 100,000x too strong, which blows out all materials

I'm not sure what you mean by this. The sun is now 100,000 Lux instead of 1 arbitrary unit. reducing the suns strength to 1 lux would remove the benefit of using physical light units altogether. 

I think the core issue you are noticing is there was a poor alignment of default settings leading to the sun being way too strong. I have adjusted the defaults, including adding a CameraAttributesPhysical to the editor default WorldEnvironment which makes the 100,000 lux sun look good by default

> * Material previews are black (for basic white material)

Fixed

> * Default camera settings create a black environment. The numbers are fine defaults, the effects are inaccurate.

In addition to the above, I also changed the background intensity to 30,000 nits so it should look good with defaults now

> * Focal length doesn't change field of view, and vica versa
> * FOV is only partially locked out under Phys Camera. Either unlock it or completely lock it.
> * Exposure settings should all be in the exposure group
> * Recommend reordering settings: Focus Distance, Focal Length, Aperture, Shutter speed, Sensitivity, Multiplier.

Done

> * Auto exposure only sometimes works, doesn't override exposure settings

I haven't changed Auto exposure in this PR, but it certainly needs a little work now.

> * The Physical Camera amount of blur is a bit weak. I was estimating the amount of blur on an object at 2m away is what 3m away should look like. So perhaps a 66.7% increase will achieve it.

Done

jcostello commented:
@clayjohn when you work on exposure, please check that the exposure `speed` affect the exposure itself. I think this is an issue in this branch

mrjustaguy commented:
With just the default omni/spot lights to light up stuff, and the default physical camera attributes, the scene is nearly pitch black right now..

jcostello commented:
Default Directional Light with Default environment, add a Practical Camera Attributes to the default environment, all pitch white.

tinmanjuggernaut commented:
> Physical Camera Attributes has a really big performance impact because the already has DOF enabled and the impact is tremendous

Can you elaborate on this statement? Is it worse than the PracticalCamera w/ DOF enabled? PhysicalCamera is supposed to have DOF by definition. A performance impact is part of the deal, but it shouldn't be worse than the other camera.

clayjohn commented:
> > Physical Camera Attributes has a really big performance impact because the already has DOF enabled and the impact is tremendous
> 
> Can you elaborate on this statement? Is it worse than the PracticalCamera w/ DOF enabled? PhysicalCamera is supposed to have DOF by definition. A performance impact is part of the deal, but it shouldn't be worse than the other camera.

With default settings the impact of DoF should also be quite low as the blur amount will be very low and the far blur will be disabled

> Default Directional Light with Default environment, add a Practical Camera Attributes to the default environment, all pitch white.

Thanks for the feedback. I have an idea to make this a little easier to work with.

Calinou commented:
> With default settings the impact of DoF should also be quite low as the blur amount will be very low and the far blur will be disabled

Should PhysicalSkyCamera (or the project settings) have a "kill switch" for DoF, especially for mobile?

clayjohn commented:
> > With default settings the impact of DoF should also be quite low as the blur amount will be very low and the far blur will be disabled
> 
> Should PhysicalSkyCamera (or the project settings) have a "kill switch" for DoF, especially for mobile?
> 
> This is worth discussing in a separate proposal, but I feel there should be a way to globally disable _demanding_ Environment features in the project settings and RenderingServer. Having to find the WorldEnvironment node feels more finnicky in comparison, and setting changes will not carry over to scenes using other WorldEnvironment nodes. This is similar to what [godotengine/godot-proposals#2455](https://github.com/godotengine/godot-proposals/issues/2455) proposes for BaseMaterial3D, but for Environment.

I think that is probably a good idea. Would need to discuss in a Proposal though.

clayjohn commented:
@tinmanjuggernaut @jcostello I just pushed an update. I think everything is fixed and finished now. Please test again and let me know if you are still running into some of the same usability issues that you were before. 

Importantly in this last push I fixed auto-exposure (I had broken something in the original rename) so it should be working the same as it does in master. I also adjusted the default CameraAttributesPractical so it defaults to an appropriate sensitivity for use in outdoor environments. Note, when using CameraAttributesPractical sensitivity is scaled so that a sensitivity of ISO100 works well for lights with brightness of 1. Finally, I renamed the setting to "rendering/lights_and_shadows/use_physical_light_units" and renamed all "rendering/shadows/*" to "rendering/lights_and_shadows/*" Physical light units is now off by default.

Calinou commented:
> I think that is probably a good idea. Would need to discuss in a Proposal though.

Done :slightly_smiling_face: https://github.com/godotengine/godot-proposals/issues/5229

jcostello commented:
Issues that I notice:

Fixes:
1) Scale issue in exposure fixed
2) Practical Camera Issue fixed

Issues:
1) Strange reflection in material (Test1.tscn)

Roughness 0
![image](https://user-images.githubusercontent.com/1168582/186057423-ca270a96-d9d2-44e6-b7b0-6e4a03e2402b.png)

Roughness 1
![image](https://user-images.githubusercontent.com/1168582/186057530-ce315f8f-3a8a-43e3-8a4f-9a27e051485e.png)

2) Reflection Probes is not working correcly (Test2.tscn)

Interior False
![image](https://user-images.githubusercontent.com/1168582/186057951-4ce2a721-dbe0-4ef7-b91d-df85b08f2577.png)

Interior True
![image](https://user-images.githubusercontent.com/1168582/186057988-b1f01b9b-5153-4ff0-a11a-95b940d5f32d.png)

MRP
[MRP.zip](https://github.com/godotengine/godot/files/9398920/MRP.zip)

clayjohn commented:
> Issues that I notice:
> 
> Fixes:
> 
> 1. Scale issue in exposure fixed
> 2. Practical Camera Issue fixed
> 
> Issues:
> 
> 1. Strange reflection in material (Test1.tscn)
> 
> Roughness 0 ![image](https://user-images.githubusercontent.com/1168582/186057423-ca270a96-d9d2-44e6-b7b0-6e4a03e2402b.png)
> 
> Roughness 1 ![image](https://user-images.githubusercontent.com/1168582/186057530-ce315f8f-3a8a-43e3-8a4f-9a27e051485e.png)
> 
> 2. Reflection Probes is not working correcly (Test2.tscn)
> 
> Interior False ![image](https://user-images.githubusercontent.com/1168582/186057951-4ce2a721-dbe0-4ef7-b91d-df85b08f2577.png)
> 
> Interior True ![image](https://user-images.githubusercontent.com/1168582/186058528-f053e6a8-2848-4f36-b588-5676d602087f.png)
> 
> MRP [MRP.zip](https://github.com/godotengine/godot/files/9398920/MRP.zip)

Thanks for testing. I am able to reproduce your second issue (with the ReflectionProbe) in master, so I won't attempt a fix here. But I am looking into the first issue now!

clayjohn commented:
@jcostello Can you try your test scene in master and see if you have the same issues? I can't reproduce your issue 1 with this PR or in master.

tinmanjuggernaut commented:
Very good. This is much better. I tested aececa3b7.

* Material previews are still black
* Does `use physical light units` have to require an editor restart?
* I would go up to f/64. Ansel Adams was known for using that f-stop, as was [this group of artists](https://en.wikipedia.org/wiki/Group_f/64).
* Default focal length is 15.65, FOV of 75. You should go back to the previous Godot default of 70 (though I think its a bit wide, Unity & UE are 60). The viewport is 70. Too wide distorts images. 75 (15.6mm focal length) is almost a fish eye.
* ISO should step to `int` or 0.1-0.01. We don't need .0001 precision. That makes it hard to slide in the inspector.
* We do need the precision for shutterspeed as it's measured down to 1/8000th of a second. The units are technically correct. But I think it makes more sense to change the unit to `1/s`. Photographers don't think in `0.001s` or `0.000125s`, they think in `1/100` or `1/8000`. UE uses `1/s` shown at the link below. I see sensitivity has an `ISO` label in the field. We could show `1/s` here too and invert it. Yes a shutterspeed could go to 8s, which would be 0.125 in this new field, but the sub 1 settings are more commonly used, so I would cater to them. Typically cameras change units, going `1/4, 1/2, 1s, 2s`. 

### Exposure
Something isn't right with the fstop exposure math. Aperture, shutterspeed and ISO are all a measurement of stops of light (brightness). A 2x increase or a 1/2 decrease doubles or halves the amount of light.
  * ISO 100 -> 200 is an increase of 1 stop, or 2x the light. 100 -> 50 is 1 stop and 1/2 the light. 
  * f/16 -> f/8 is 1 stop.
  * 1/100 -> 1/200 is 1 stop. 
  
I started at the defaults: f/16, 1/100 shutter, 100 iso. I should be able to go from f/16 to f8 and see the brightness increase to 2x, then decrease shutterspeed from 1/100 to 1/200 and see the brightness return to the original level. ISO and shutterspeed correctly inversely correlate. Halving fstop and shutterspeed or ISO results in 2x brightness. 1/4th results in 4x brightness. All of the viewports below should have the exact same exposure.

I came across this forumula on a [UE manual page](https://docs.unrealengine.com/5.0/en-US/auto-exposure-in-unreal-engine/) `EV100 = log2(Aperture^2 / Shutter Speed * 100/ISO)`. This page also shows shutter speed 1/s and a whole lot of other information that is worth looking at.

![image](https://user-images.githubusercontent.com/632766/186439932-2f14a301-1757-446e-aba0-4615afbbedfa.png)

* Autoexposure still doesn't work in realtime adjustment of camera, and only sometimes once enabled. 
* If exposure is way too hot, auto exposure won't work at all (defaults + fstop/2). 
* What would be good is if AE was enabled, I could adjust aperture in realtime and have it automatically adjust shutterspeed and ISO in realtime, as a camera works in A mode (P full auto, Aperture, Shutter, Manual). I understand if it's out of scope.


### Blur

* Focus distance < 0.1 reverses the blur effect (90mm, f/4). At 0.01, it's gone, and completely in focus . If it's a math thing, I'd clamp it to 0.1 or 10cm. That's fine. All physical lenses have a minimum focus distance. Nikon 60mm is 18.5cm, Nikon 105mm 31.4cm, Nikon 200m is 48.8cm.

* The focal planes (hex/low) move pixels of a focused object up and to the left significantly enough to notice and be disconcerting. It affects facial details and edges. Racking focus, my objects pops up then back down as focus hits them. It looks like it's a directional blur is used instead of a radial blur. It only moves pixels when close to the focus plane (eg 2.4-4 for an object @ 3). Outside of that range they don't move. Hex/high is less pronounced, but it may be because the blur amount is different (see below).

https://user-images.githubusercontent.com/632766/186451072-b323c0af-e4b7-455b-be89-9fa2e6aaca1e.mp4

![moving focus](https://user-images.githubusercontent.com/632766/186449614-2871d539-7cf0-46f3-952f-a99de07004a7.gif)

* I compared the blur 90mm with the dofsimulator site and my experience, and I think the blur amount looks very good (on the default low quality, hexagon). 
* Is the previous blur quality now a project setting called `depth_of_field_bokeh_quality`? 
* `rendering/camera/depth_of_field/depth_of_field_bokeh_quality` How about reducing one to `dof_`?
* The blur amount varies between the levels:
  * Hexagon/Low is pretty accurate. 
  * Hexagon/Very low is the same as low. 
  * Hexagon/Medium is about 1/4-1/2 the blur it should be
  * Hexagon/High is the same as medium.
  * Box is the same as above
  * Circle, all levels are consistent and accurate
  * Jitter doesn't seem to do anything. I see no effect on the highest circle or lowest box.

* Circle is not usable for a racking focus camera. VLow the blur is unstable and dances around. High looks amazing, but it's very slow and produces artifacts on edges of objects coming into focus:

https://user-images.githubusercontent.com/632766/186460824-8e273f1b-0a9e-41b7-8750-2c299c79d47e.mp4

* Hexagon is the best for a racking focus camera. If the blur levels are made consistent and pixel movement is fixed, this is what I'll use. Circle has too many issues. Box is less attractive and around the same speed as hexagon, so no reason to use it.

Nice work so far!

Calinou commented:
> Default focal length is 15.65, FOV of 75. You should go back to the previous Godot default of 70 (though I think its a bit wide, Unity & UE are 60). The viewport is 70. Too wide distorts images. 75 (15.6mm focal length) is almost a fish eye.

For reference, a 90° horizontal FOV on a 4:3 display is a vertical FOV of 73.74°. This is generally considered a good baseline for desktop games. 75° is closer to that compared to 70°, so I prefer it as a PC gamer :slightly_smiling_face:

tinmanjuggernaut commented:
The issue with wide FOV is distortion. A 50mm lens is widely considered equivalent to how human vision works so is called a normal lens and the model looks normal. 24mm-35mm is wide angle, and below 24 is ultra wide angle.

![image](https://user-images.githubusercontent.com/632766/186492932-aa805671-f9d6-4ee5-8ea4-b8c5f4c61989.png)

75 FOV / 15.65mm is an extreme setting and should not be a default. Even 70 is extreme, but going higher is worse. If developers or game players want to use 75 or higher, for their personal preference fine, but for a default for a general usage engine, no more than 70 for default, and I wouldn't even do that if building the engine from scratch. 

The type of game may determine FOV, whether FPS, TPS, etc. Since the engine is general purpose it should default to a normal level so people's games aren't distorted (or less distorted). Recall how many projects came out blue because Godot's sky defaults were extreme and people never changed the settings as I demonstrated in my proposal.

> [73-75] as to match what most PC games are doing by default

Which are these "most PC games"? Are they TPS or FPS? Are you sure the FOVs you are looking at vertical or horizontal? Many do not specify, and it can be difficult to tell when playing or viewing game footage. Plus many games dynamically change their FOV, such as in modes where the player focuses, zooms, looks in a scope, or in dialog. 

A random sampling of popular games FOVs from https://www.pcgamingwiki.com/wiki/ shows FOVs all over the place if they are fixed, and the ranges are not standardized either. Even games in the same series are different. Here are engines plus a random sampling.
Godot: 70 vertical current release, and viewport
Unity: 60 vertical
UE4: 59 vertical (specified as 90 horizontal @ 16:9)
Witcher 3: 60 vertical
Horizon Zero Dawn: 70-100 (could be either)
God of War 4: 73 (probably horizontal)
GTAV: 55 vertical
Call of Duty Black ops: 65-80 vertical
Call of Duty warzone: 36-89 vertical
Doom Eternal: 90-120 (probably horizontal)
Halo Infinite: 65-120, 78 default (dev hints vertical, but could be horizontal)

clayjohn commented:
@tinmanjuggernaut Thank you for the detailed feedback, as always it is super helpful to have a user's perspective

> * Material previews are still black

Should be fixed now!

> * Does `use physical light units` have to require an editor restart?

Unfortunately yes. Because light units can now range from 0.01 to over 100,000, we run into issues with floating point precision so we have to bake exposure into light units before rendering. While this isn't an outright bar to changing the setting without restart it would add significant complexity. The real bar is exposing appropriate settings and properties when physical light units are enabled. The editor has to know what the setting is when determining what units/properties are exposed and used.

> * I would go up to f/64. Ansel Adams was known for using that f-stop, as was [this group of artists](https://en.wikipedia.org/wiki/Group_f/64).

Done.

> * Default focal length is 15.65, FOV of 75. You should go back to the previous Godot default of 70 (though I think its a bit wide, Unity & UE are 60). The viewport is 70. Too wide distorts images. 75 (15.6mm focal length) is almost a fish eye.

Right now it matches the Godot default of 75. I am not opposed to having a different default for physical attributes than the default camera, but that would require a discussion with others. I'll gather some feedback and return to this. 

> * ISO should step to `int` or 0.1-0.01. We don't need .0001 precision. That makes it hard to slide in the inspector.

Changed to step 0.1, default is back to ISO 100. Now the default setting should match CameraAttributesPhysical at aperture 16 and shutter speed 1/100.

> * We do need the precision for shutterspeed as it's measured down to 1/8000th of a second. The units are technically correct. But I think it makes more sense to change the unit to `1/s`. Photographers don't think in `0.001s` or `0.000125s`, they think in `1/100` or `1/8000`. UE uses `1/s` shown at the link below. I see sensitivity has an `ISO` label in the field. We could show `1/s` here too and invert it. Yes a shutterspeed could go to 8s, which would be 0.125 in this new field, but the sub 1 settings are more commonly used, so I would cater to them. Typically cameras change units, going `1/4, 1/2, 1s, 2s`.

That makes sense, changed. 

> ### Exposure
> Something isn't right with the fstop exposure math. 

I think you are mixing up EV "stops" and f-stops. The relationship between f-stop and EV is that a decrease of f-stop by half results in a 4 times decrease in brightness (e.g. 2 EV stops) while a 2 times increase in shutter speed only compensates by increasing the EV by 1. In other words, if you cut your f-stop in half, you need to increase your shutter speed 4x (when measured in 1/s). Wikipedia has a helpful table to illustrate https://en.wikipedia.org/wiki/Exposure_value#EV_as_an_indicator_of_camera_settings

> * Autoexposure still doesn't work in realtime adjustment of camera, and only sometimes once enabled.
> * If exposure is way too hot, auto exposure won't work at all (defaults + fstop/2).

I'll look into this. I am trying to avoid overhauling auto-exposure as this PR is getting way too large already, but this sounds like an issue worth fixing. 

> * What would be good is if AE was enabled, I could adjust aperture in realtime and have it automatically adjust shutterspeed and ISO in realtime, as a camera works in A mode (P full auto, Aperture, Shutter, Manual). I understand if it's out of scope.

I considered this, some engines do let users select between aperture priority mode, shutter speed priority mode, and aperture and shutter priority mode (only adjust sensitivity). In practice, I am not sure this is useful in real games as changing sensitivity should be enough and eventually shutter speed will impact motion blur (similar to how aperture affects DoF).

> ### Blur
> * Focus distance < 0.1 reverses the blur effect (90mm, f/4). At 0.01, it's gone, and completely in focus . If it's a math thing, I'd clamp it to 0.1 or 10cm. That's fine. All physical lenses have a minimum focus distance. Nikon 60mm is 18.5cm, Nikon 105mm 31.4cm, Nikon 200m is 48.8cm.

I'll take a look, I'm not sure why this is happening. But if a minimum focus distance is an acceptable fix then this shouldn't be too much work to address.

> * The focal planes (hex/low) move pixels of a focused object up and to the left significantly enough to notice and be disconcerting. It affects facial details and edges. Racking focus, my objects pops up then back down as focus hits them. It looks like it's a directional blur is used instead of a radial blur. It only moves pixels when close to the focus plane (eg 2.4-4 for an object @ 3). Outside of that range they don't move. Hex/high is less pronounced, but it may be because the blur amount is different (see below).

I would open this as a separate issue. This is likely an issue with the downsampling we do before calculating DoF blur and it is out of scope for this PR. Nevertheless, it should not be too hard to fix if I am right. 

> * Is the previous blur quality now a project setting called `depth_of_field_bokeh_quality`?
> * `rendering/camera/depth_of_field/depth_of_field_bokeh_quality` How about reducing one to `dof_`?
> * The blur amount varies between the levels:
>   
>   * Hexagon/Low is pretty accurate.
>   * Hexagon/Very low is the same as low.
>   * Hexagon/Medium is about 1/4-1/2 the blur it should be
>   * Hexagon/High is the same as medium.
>   * Box is the same as above
>   * Circle, all levels are consistent and accurate
>   * Jitter doesn't seem to do anything. I see no effect on any bokeh or quality level.
> * Circle is not usable for a racking focus camera. VLow the blur is unstable and dances around. High looks amazing, but it's very slow and produces artifacts on edges of objects coming into focus:
> 
> * Hexagon is the best for a racking focus camera. If the blur levels are made consistent and pixel movement is fixed, this is what I'll use. Circle has too many issues. I might use it in static scenes, creating screenshots, or dialog scenes. Box is less attractive and around the same speed as hexagon, so no reason to use it.

I noticed a lot of these issues with DoF blur when testing as well. Some should be easy fixes (the blur amount changing with quality settings for example). But overhauling DoF is out of scope for this PR. If you have the time, please open a tracker issue to report the issues you have with DoF, if you don't have time, I can do it once I finalize this PR. 

Again, thank you very much for your testing and feedback

clayjohn commented:
> Focus distance < 0.1 reverses the blur effect (90mm, f/4). At 0.01, it's gone, and completely in focus . If it's a math thing, I'd clamp it to 0.1 or 10cm. That's fine. All physical lenses have a minimum focus distance. Nikon 60mm is 18.5cm, Nikon 105mm 31.4cm, Nikon 200m is 48.8cm.

Looking into this closer, the effect reverses if the focus distance is less than the focal length. I will just clamp focus distance to focal length to avoid this.

tinmanjuggernaut commented:
> I think you are mixing up EV "stops" and f-stops. The relationship between f-stop and EV is that a decrease of f-stop by half results in a 4 times decrease in brightness (e.g. 2 EV stops) while a 2 times increase in shutter speed only compensates by increasing the EV by 1.

You're right, I was mistaken. I forgot about the f-stops in between the powers of two. Each stop is double or half light. However f/8 to f/4 is 2 stops. f/8 to f/5.6 is 1 stop as your table shows.

clayjohn commented:
@tinmanjuggernaut Updated to fix auto exposure issues and changed default focal length to 35mm. Barring the bugs with DoF, I think I have addressed all of your reported issues. I would appreciate if you gave it another try!

tinmanjuggernaut commented:
* I didn't see a default of 35mm. But whatever you guys decide is fine, I said my piece.

* You could lower the aperture range to say 0.5, in case someone wants to get just a bit of extra blur, as such lenses exist [f/0.95](https://www.kenrockwell.com/nikon/f-095.htm), [f/0.7](https://en.wikipedia.org/wiki/Carl_Zeiss_Planar_50mm_f/0.7), and [the lowest for spherical elements is f/0.5, or f/0.001? for aspheric lenses](https://www.dpreview.com/forums/post/16805340). In practice, I've only seen f/0.95 lenses and didn't know about the others until I looked it up.

* Materials are broken on the ends of the ranges, eg. Everything that is overexposed should stay white.
  * f/4, shutter 0.1, iso 100 = white sky, white materials
  * f/3.5, shutter 0.1, iso 100 => white sky, materials are black !!
  * f/1, shutter 0.1-1.271 => white sky, materials are black !!
  * f/1, shutter 1.272-8000 => all white to color
  * f/1, shutter 0.1, iso 0.1-7.8 => all white
  * f/1, shutter 0.1, iso 7.9-32000 => white sky, black materials !!

* I'm not sure what auto exposure issues you fixed, because it still looks broken. I made the image very overexposed, all white, turned on auto exposure and it darkened the image but left it still very over exposed. Then I mildly under exposed (f/5, shutter 6000, iso 100), it brightens but its still under exposed. Then defaults (f/16, sh100, iso 100) turned on auto exposure and it did properly expose the middle of the image, but it left part of the image very blown out. I understand you have another PR for overhauling this.

![image](https://user-images.githubusercontent.com/632766/186773818-2bcc2917-c9b5-4911-9097-2f31c401d00e.png)


Otherwise, it looks good. I'll make an issue for the blur issues. Thanks!

jcostello commented:
> @jcostello Can you try your test scene in master and see if you have the same issues? I can't reproduce your issue 1 with this PR or in master.

Seems to be an issue with master

clayjohn commented:
> I'm not sure what auto exposure issues you fixed, because it still looks broken. I made the image very overexposed, all white, turned on auto exposure and it darkened the image but left it still very over exposed. Then I mildly under exposed (f/5, shutter 6000, iso 100), it brightens but its still under exposed. Then defaults (f/16, sh100, iso 100) turned on auto exposure and it did properly expose the middle of the image, but it left part of the image very blown out. I understand you have another PR for overhauling this.

@tinmanjuggernaut What settings are you using for auto-exposure? You may have to expand the luminance range. We set caps on the luminance range in auto exposure so that you can still create bright and dark areas in your game Matius Goldberg explains it [really well i the original PR](https://github.com/godotengine/godot-proposals/issues/4257#issuecomment-1075484745). The default settings work will for default exposure ranges, if you push exposure to the extreme you will need to adjust the auto exposure to accept extreme values as well.

tinmanjuggernaut commented:
>> f/1, shutter 0.1-1.271 => white sky, materials are black !!
>> f/1, shutter 1.272-8000 => all white to color
> I wasn't able to reproduce this on my machine (intel integrated). I wonder if you are getting NaNs when your overexposed by a large enough factor.

NVidia RTX 3070. Exposure already allows many stops of blow out before turning black. Perhaps we can determine what exposure value causes white to turn to black and clamp it. Something like 10 stops over exposed should be sufficient, as anything over say 5 is going to be all white anyway.

> What settings are you using for auto-exposure? 

I just left the defaults. When I and all other photographers and cinematographers enable auto exposure, the expectation is that the camera will adjust exposure of the current image to 50% middle gray. What "50% middle gray" means depends on the metering method, which could take a straight average of all pixels, or it could weight the center more heavily, or use a histogram, or many other methods. 

Any settings for auto exposure are to override it, meaning auto expose to 50% gray, then add or minus stops. I would expect AE settings to have only: 
* AE Metering mode (full image, spot, histogram, etc)
* Increase/decrease exposure (after AE) by 5-10 stops
* AE application speed

What I experienced is:
* Way overexpose the image, and AE darkened it slightly, but still way overexposed. 
* Underexpose the image, AE slightly brightened it, but left it underexposed.
* At other times over or underexposed, AE properly exposed it.

At the moment it is unreliable, inconsistent, not intuitive, and not how real analog or digital cameras work.

BastiaanOlij commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -1083,43 +1085,62 @@ int RendererSceneRenderRD::get_directional_light_shadow_size(RID p_light_intance
 
-RID RendererSceneRenderRD::camera_effects_allocate() {
-	return camera_effects_owner.allocate_rid();
+RID RendererSceneRenderRD::camera_attributes_allocate() {
```
Should we move this into one of our storage objects? Or create a Camera attributes storage object?

clayjohn replied:
I created a CameraAttributes storage object that is shared by the various backends

BastiaanOlij commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -1890,7 +1911,7 @@ void RendererSceneRenderRD::_render_buffers_post_process_and_tonemap(const Rende
 				float z_near = p_render_data->view_projection[i].get_z_near();
 				float z_far = p_render_data->view_projection[i].get_z_far();
-				bokeh_dof->bokeh_dof_compute(buffers, camfx->dof_blur_far_enabled, camfx->dof_blur_far_distance, camfx->dof_blur_far_transition, camfx->dof_blur_near_enabled, camfx->dof_blur_near_distance, camfx->dof_blur_near_transition, bokeh_size, dof_blur_bokeh_shape, dof_blur_quality, dof_blur_use_jitter, z_near, z_far, p_render_data->cam_orthogonal);
+				bokeh_dof->bokeh_dof_compute(buffers, cam_attributes->dof_blur_far_enabled, cam_attributes->dof_blur_far_distance, cam_attributes->dof_blur_far_transition, cam_attributes->dof_blur_near_enabled, cam_attributes->dof_blur_near_distance, cam_attributes->dof_blur_near_transition, bokeh_size, dof_blur_bokeh_shape, dof_blur_quality, dof_blur_use_jitter, z_near, z_far, p_render_data->cam_orthogonal) ...
```
Nitpicking but should we just give the cam_attributes pointer to bokeh_dof_compute?

clayjohn replied:
Done

BastiaanOlij reviewed (approved):
As far as I can judge code wise this all looks pretty solid to me. I must admit not being fully up to speed with the in's and outs of physical units but from what I understand of the general idea, I really like the approach of having separate settings for practical and physical units.

Nitpicking for the most part but my only two things I'd like to have further discussion on:
- Do the physical units result in higher intensity colors being stored in our 3D color buffer and thus would we run into trouble with the limited precision on the mobile renderer?
- Should we move `CameraAttributes` within the render server into its own storage object, having it in `RenderSceneRender` seems counter productive in light of all the other work we've been doing on this front recently (off course keeping in mind the older `CameraEffects` was already in `RenderSceneRender`, so this was really already outstanding).

jcostello commented:
@clayjohn Models in the import preview modal look all white (probably missing the camera attributes there)

clayjohn commented:
> What I experienced is:
> 
> * Way overexpose the image, and AE darkened it slightly, but still way overexposed.
> * Underexpose the image, AE slightly brightened it, but left it underexposed.
> * At other times over or underexposed, AE properly exposed it.
> 
> At the moment it is unreliable, inconsistent, not intuitive, and not how real analog or digital cameras work.

I have expanded the default range to handle extreme values better. It may be better to expose the range in EV-stops instead of luminance as EV scales better to extreme values. I'll try it out and see how it goes

> @clayjohn Models in the import preview modal look all white (probably missing the camera attributes there)

Good catch! Fixed.

> Do the physical units result in higher intensity colors being stored in our 3D color buffer and thus would we run into trouble with the limited precision on the mobile renderer?

Not necessarily, light values are always pre-exposed so that the value stored in the buffer is approximately the same as it was before (for a correctly exposed scene). This goes for the 3D color buffer, for IBL, for Lightmaps, for VoxelGI, for SDFGI etc.

clayjohn commented:
> Materials are broken on the ends of the ranges, eg. Everything that is overexposed should stay white.
> 
> f/4, shutter 0.1, iso 100 = white sky, white materials
> f/3.5, shutter 0.1, iso 100 => white sky, materials are black !!
> f/1, shutter 0.1-1.271 => white sky, materials are black !!
> f/1, shutter 1.272-8000 => all white to color
> f/1, shutter 0.1, iso 0.1-7.8 => all white
> f/1, shutter 0.1, iso 7.9-32000 => white sky, black materials !!

I just figured this one out and it is now fixed!

Rebased to fix merge conflicts, this should now be ready to go!

reduz commented:
This looks super solid to me, incredible work!!

akien-mga commented:
Let's merge! Thanks Clay for the awesome work, and thanks to everyone involved in testing and reporting bugs.

Once merged, let's open dedicated issues for further bugs.
